File: app/src/main/java/com/winlator/cmod/core/PreloaderDialog.java
package com.winlator.cmod.core;

import android.app.Activity;
import android.app.Dialog;
import android.graphics.Bitmap;
import android.os.Handler;
import android.os.Looper;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.widget.ImageView;
import android.widget.TextView;

import com.winlator.cmod.R;

public class PreloaderDialog {
    private final Activity activity;
    private Dialog dialog;
    private final Handler dotsHandler = new Handler(Looper.getMainLooper());
    private String baseStatus = "";
    private int dotsState = 0;
    private final Runnable dotsRunnable = new Runnable() {
        @Override
        public void run() {
            if (dialog == null || !dialog.isShowing()) return;
            TextView tv = dialog.findViewById(R.id.TextView);
            if (tv != null) {
                StringBuilder sb = new StringBuilder(baseStatus);
                for (int i = 0; i < dotsState; i++) sb.append('.');
                tv.setText(sb.toString());
            }
            dotsState = (dotsState + 1) % 4;
            dotsHandler.postDelayed(this, 400);
        }
    };

    public PreloaderDialog(Activity activity) {
        this.activity = activity;
    }

    private void create() {
        if (dialog != null) return;
        dialog = new Dialog(activity, android.R.style.Theme_Translucent_NoTitleBar_Fullscreen);
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);
        dialog.setCancelable(false);
        dialog.setCanceledOnTouchOutside(false);
        dialog.setContentView(R.layout.preloader_dialog);

        Window window = dialog.getWindow();
        if (window != null) {
            window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE);
            window.clearFlags(WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE);
        }
    }

    public synchronized void show(int textResId) {
        if (isShowing()) return;
        close();
        if (dialog == null) create();
        baseStatus = stripTrailingDots(activity.getString(textResId));
        ((TextView)dialog.findViewById(R.id.TextView)).setText(baseStatus);
        applyPendingTexts();
        dialog.show();
        startFadeIn();
        startDotsAnimation();
    }

    private CharSequence pendingTitle = null;
    private CharSequence pendingSubtitle = null;

    public synchronized void setTitle(CharSequence title) {
        this.pendingTitle = title;
        if (dialog != null) {
            TextView titleView = dialog.findViewById(R.id.TitleTextView);
            if (titleView != null && title != null) titleView.setText(title);
        }
    }

    public synchronized void setSubtitle(CharSequence subtitle) {
        this.pendingSubtitle = subtitle;
        if (dialog != null) {
            TextView subView = dialog.findViewById(R.id.SubtitleTextView);
            if (subView != null) {
                subView.setText(subtitle == null ? "" : subtitle);
                subView.setVisibility(subtitle == null || subtitle.length() == 0 ? View.GONE : View.VISIBLE);
            }
        }
    }

    public void setSubtitleOnUiThread(final CharSequence subtitle) {
        activity.runOnUiThread(() -> setSubtitle(subtitle));
    }

    public void setStageOnUiThread(final CharSequence stage) {
        activity.runOnUiThread(() -> setStage(stage));
    }

    public synchronized void setStage(CharSequence stage) {
        baseStatus = stripTrailingDots(stage == null ? "" : stage.toString());
        dotsState = 0;
        if (dialog != null && isShowing()) {
            TextView tv = dialog.findViewById(R.id.TextView);
            if (tv != null) tv.setText(baseStatus);
        }
    }

    private void applyPendingTexts() {
        if (dialog == null) return;
        if (pendingTitle != null) {
            TextView titleView = dialog.findViewById(R.id.TitleTextView);
            if (titleView != null) titleView.setText(pendingTitle);
        }
        TextView subView = dialog.findViewById(R.id.SubtitleTextView);
        if (subView != null) {
            if (pendingSubtitle != null && pendingSubtitle.length() > 0) {
                subView.setText(pendingSubtitle);
                subView.setVisibility(View.VISIBLE);
            } else {
                subView.setVisibility(View.GONE);
            }
        }
    }

    private void startFadeIn() {
        if (dialog == null) return;
        View root = dialog.findViewById(R.id.PreloaderRoot);
        if (root == null) return;
        root.setAlpha(0f);
        root.animate().alpha(1f).setDuration(280).setInterpolator(new AccelerateDecelerateInterpolator()).start();
    }

    private void startDotsAnimation() {
        dotsHandler.removeCallbacks(dotsRunnable);
        dotsState = 0;
        dotsHandler.postDelayed(dotsRunnable, 400);
    }

    private static String stripTrailingDots(String s) {
        if (s == null) return "";
        int end = s.length();
        while (end > 0) {
            char c = s.charAt(end - 1);
            if (c == '.' || c == '\u2026' || c == ' ') end--;
            else break;
        }
        return s.substring(0, end);
    }

    public void showOnUiThread(final int textResId) {
        activity.runOnUiThread(() -> show(textResId));
    }
    
    /**
     * Обновить текст в уже открытом диалоге
     */
    public synchronized void updateText(String text) {
        setStage(text);
    }
    
    /**
     * Обновить текст в диалоге из UI потока
     */
    public void updateTextOnUiThread(final String text) {
        activity.runOnUiThread(() -> updateText(text));
    }

    public synchronized void close() {
        dotsHandler.removeCallbacks(dotsRunnable);
        if (dialog == null) return;
        final Dialog d = dialog;
        dialog = null;
        try {
            View root = d.findViewById(R.id.PreloaderRoot);
            if (root != null && d.isShowing()) {
                root.animate().alpha(0f).setDuration(260)
                    .setInterpolator(new AccelerateDecelerateInterpolator())
                    .withEndAction(() -> { try { d.dismiss(); } catch (Exception ignored) {} })
                    .start();
            } else {
                d.dismiss();
            }
        } catch (Exception e) {
            try { d.dismiss(); } catch (Exception ignored) {}
        }
    }

    public synchronized void setCoverArt(Bitmap bitmap) {
        if (dialog == null) return;
        ImageView coverView = dialog.findViewById(R.id.CoverArtBackground);
        View scrim = dialog.findViewById(R.id.CoverArtScrim);
        if (coverView == null || scrim == null) return;
        if (bitmap != null) {
            coverView.setImageBitmap(bitmap);
            coverView.setVisibility(View.VISIBLE);
            scrim.setVisibility(View.VISIBLE);
        } else {
            coverView.setVisibility(View.GONE);
            scrim.setVisibility(View.GONE);
        }
    }

    public void setCoverArtOnUiThread(final Bitmap bitmap) {
        activity.runOnUiThread(() -> setCoverArt(bitmap));
    }

    public void closeOnUiThread() {
        activity.runOnUiThread(this::close);
    }

    public boolean isShowing() {
        return dialog != null && dialog.isShowing();
    }
}

File: app/src/main/java/com/winlator/cmod/core/PreloaderDialog.kt
package com.winlator.cmod.core

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.graphics.Bitmap
import android.hardware.input.InputManager
import android.os.Handler
import android.os.Looper
import android.view.View
import androidx.compose.animation.*
import androidx.compose.animation.core.*
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.Computer
import androidx.compose.material.icons.filled.SportsEsports
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.blur
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Shadow
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.unit.IntOffset
import com.winlator.cmod.R
import com.winlator.cmod.inputcontrols.ExternalController
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlin.math.roundToInt
import coil.compose.AsyncImage

class PreloaderDialog(private val activity: Activity) {
    private var dialog: Dialog? = null

    private var isVisible by mutableStateOf(false)
    private var titleText by mutableStateOf<CharSequence?>(null)
    private var subtitleText by mutableStateOf<CharSequence?>(null)
    private var stageText by mutableStateOf<CharSequence?>(null)
    private var coverArtBitmap by mutableStateOf<Bitmap?>(null)
    private var _steamAppId by mutableStateOf<String?>(null)

    private var exitTriggered by mutableStateOf(false)
    private var onFadeOutComplete: (() -> Unit)? = null

    private fun create() {
        if (dialog != null) return
        dialog = Dialog(activity, android.R.style.Theme_Translucent_NoTitleBar_Fullscreen).apply {
            requestWindowFeature(android.view.Window.FEATURE_NO_TITLE)
            setCancelable(false)
            setCanceledOnTouchOutside(false)

            val composeView = ComposeView(activity).apply {
                setContent {
                    val isDark = com.winlator.cmod.contentdialog.ContentDialog.shouldUseDarkDialog(activity)
                    com.winlator.cmod.ui.theme.WinlatorTheme(darkTheme = isDark) {
                        SteamDeckBootScreen()
                    }
                }
            }
            setContentView(composeView)

            window?.let { win ->
                win.clearFlags(android.view.WindowManager.LayoutParams.FLAG_NOT_TOUCHABLE)
                win.clearFlags(android.view.WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE)

                // Полный экран: скрыть статус-бар и навигационный бар
                win.addFlags(android.view.WindowManager.LayoutParams.FLAG_FULLSCREEN)
                win.addFlags(android.view.WindowManager.LayoutParams.FLAG_LAYOUT_NO_LIMITS)
                win.addFlags(android.view.WindowManager.LayoutParams.FLAG_LAYOUT_IN_SCREEN)

                // Контент под системными барами, immersive
                win.decorView.systemUiVisibility =
                    android.view.View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY or
                    android.view.View.SYSTEM_UI_FLAG_FULLSCREEN or
                    android.view.View.SYSTEM_UI_FLAG_HIDE_NAVIGATION or
                    android.view.View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                    android.view.View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN or
                    android.view.View.SYSTEM_UI_FLAG_LAYOUT_HIDE_NAVIGATION

                // Покрыть вырез камеры (notch) контентом
                if (android.os.Build.VERSION.SDK_INT >= 28) {
                    val lp = win.attributes
                    lp.layoutInDisplayCutoutMode =
                        android.view.WindowManager.LayoutParams.LAYOUT_IN_DISPLAY_CUTOUT_MODE_SHORT_EDGES
                    win.attributes = lp
                }
            }

            try {
                val viewClass = Class.forName("android.view.View")
                val viewTreeLifecycleOwnerClass = Class.forName("androidx.lifecycle.ViewTreeLifecycleOwner")
                val setLifecycleMethod = viewTreeLifecycleOwnerClass.getMethod("set", viewClass, Class.forName("androidx.lifecycle.LifecycleOwner"))
                val viewTreeViewModelStoreOwnerClass = Class.forName("androidx.lifecycle.ViewTreeViewModelStoreOwner")
                val setViewModelStoreMethod = viewTreeViewModelStoreOwnerClass.getMethod("set", viewClass, Class.forName("androidx.lifecycle.ViewModelStoreOwner"))
                val viewTreeSavedStateRegistryOwnerClass = Class.forName("androidx.savedstate.ViewTreeSavedStateRegistryOwner")
                val setSavedStateMethod = viewTreeSavedStateRegistryOwnerClass.getMethod("set", viewClass, Class.forName("androidx.savedstate.SavedStateRegistryOwner"))

                val setOwners = { view: View ->
                    setLifecycleMethod.invoke(null, view, activity)
                    setViewModelStoreMethod.invoke(null, view, activity)
                    setSavedStateMethod.invoke(null, view, activity)
                }

                window?.decorView?.let { setOwners(it) }
                setOwners(composeView)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    @Composable
    private fun SteamDeckBootScreen() {
        var visible by remember { mutableStateOf(false) }

        LaunchedEffect(Unit) {
            visible = true
        }

        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color(0xFF0F1016))
        ) {
            val appId = _steamAppId?.toIntOrNull()

            // Subtle blurred background (Steam Hero / local cover) — не перетягивает внимание
            if (appId != null && appId > 0) {
                val heroUrl = "https://shared.fastly.steamstatic.com/store_item_assets/steam/apps/$appId/library_hero.jpg"
                val capsuleUrl = "https://shared.fastly.steamstatic.com/store_item_assets/steam/apps/$appId/capsule_616x353.jpg"
                var heroFile by remember(appId) { mutableStateOf<java.io.File?>(SteamImageCache.getCachedFile(activity, heroUrl)) }

                LaunchedEffect(appId) {
                    if (heroFile == null) {
                        heroFile = SteamImageCache.downloadIfNeeded(activity, heroUrl)
                            ?: SteamImageCache.downloadIfNeeded(activity, capsuleUrl)
                    }
                }

                heroFile?.let { file ->
                    AsyncImage(
                        model = file,
                        contentDescription = null,
                        modifier = Modifier
                            .fillMaxSize()
                            .blur(48.dp)
                            .alpha(0.35f),
                        contentScale = ContentScale.Crop
                    )
                }
            } else {
                coverArtBitmap?.let { bmp ->
                    Image(
                        bitmap = bmp.asImageBitmap(),
                        contentDescription = null,
                        contentScale = ContentScale.Crop,
                        modifier = Modifier
                            .fillMaxSize()
                            .blur(48.dp)
                            .alpha(0.35f)
                    )
                }
            }

            // Радиальный виньетный градиент (центр ярче, края темнее) — фокус на анимации
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .background(
                        Brush.radialGradient(
                            colors = listOf(
                                Color(0x00000000),
                                Color(0x66000000),
                                Color(0xCC0F1016)
                            ),
                            radius = 1200f
                        )
                    )
            )

            // Уведомление о контроллере — плавно выплывает справа сверху
            ControllerHintToast(
                modifier = Modifier
                    .align(Alignment.TopEnd)
                    .padding(top = 32.dp, end = 20.dp)
            )

            // Центральная композиция (иконка Windows → стрелка → миниатюра обложки)
            AnimatedVisibility(
                visible = visible,
                enter = fadeIn(tween(600)) + scaleIn(initialScale = 0.92f, animationSpec = tween(600, easing = EaseOutCubic)),
                modifier = Modifier.align(Alignment.Center)
            ) {
                Box(
                    modifier = Modifier.fillMaxSize(),
                    contentAlignment = Alignment.Center
                ) {
                    BootCenterRow(
                        bitmap = coverArtBitmap,
                        appId = appId,
                        activity = activity
                    )
                }
            }

            // Нижняя панель: полоска прогресса + стадия
            Column(
                modifier = Modifier
                    .align(Alignment.BottomCenter)
                    .fillMaxWidth()
                    .padding(horizontal = 32.dp, vertical = 56.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                val displayStage = stageText?.toString() ?: ""

                AnimatedContent(
                    targetState = displayStage,
                    transitionSpec = {
                        (fadeIn(tween(280)) + slideInVertically(animationSpec = tween(280)) { it / 4 })
                            .togetherWith(fadeOut(tween(180)) + slideOutVertically(animationSpec = tween(180)) { -it / 4 })
                    },
                    label = "stage"
                ) { stage ->
                    if (stage.isNotEmpty()) {
                        Text(
                            text = stage,
                            style = MaterialTheme.typography.bodyLarge.copy(
                                fontWeight = FontWeight.SemiBold,
                                shadow = Shadow(
                                    color = Color.Black.copy(alpha = 0.85f),
                                    offset = Offset(1f, 2f),
                                    blurRadius = 4f
                                )
                            ),
                            color = Color.White.copy(alpha = 0.85f),
                            letterSpacing = 0.4.sp,
                            textAlign = TextAlign.Center
                        )
                    } else {
                        // Резервируем высоту, чтобы колонка не «прыгала» при первой стадии
                        Spacer(modifier = Modifier.height(24.dp))
                    }
                }

                Spacer(modifier = Modifier.height(18.dp))

                LoadingBar()
            }
        }
    }

    /**
     * Плавающее уведомление в правом верхнем углу: какой контроллер используется.
     * Если подключён физический геймпад — показывает его модель.
     * Если нет — показывает «Виртуальные кнопки».
     */
    @Composable
    private fun ControllerHintToast(modifier: Modifier = Modifier) {
        val context = LocalContext.current
        var controllerName by remember { mutableStateOf<String?>(null) }

        DisposableEffect(Unit) {
            val inputManager = context.getSystemService(Context.INPUT_SERVICE) as InputManager
            val refresh = {
                controllerName = ExternalController.getControllers().firstOrNull()?.name
            }
            val listener = object : InputManager.InputDeviceListener {
                override fun onInputDeviceAdded(deviceId: Int) = refresh()
                override fun onInputDeviceRemoved(deviceId: Int) = refresh()
                override fun onInputDeviceChanged(deviceId: Int) = refresh()
            }
            refresh()
            inputManager.registerInputDeviceListener(listener, Handler(Looper.getMainLooper()))
            onDispose { inputManager.unregisterInputDeviceListener(listener) }
        }

        val show = controllerName != null
        val text = controllerName ?: context.getString(R.string.controller_virtual)

        // Плавный выезд справа: стартуем за экраном, после появления диалога выезжаем
        val slide = remember { Animatable(1f) }
        val fade = remember { Animatable(0f) }
        LaunchedEffect(Unit) {
            delay(400)
            launch { slide.animateTo(0f, tween(450, easing = EaseOutCubic)) }
            launch { fade.animateTo(1f, tween(450, easing = EaseOutCubic)) }
        }

        Surface(
            shape = RoundedCornerShape(14.dp),
            color = Color.Black.copy(alpha = 0.55f),
            shadowElevation = 10.dp,
            modifier = modifier.graphicsLayer {
                translationX = slide.value * 240.dp.toPx()
                alpha = fade.value
            }
        ) {
            Row(
                    modifier = Modifier.padding(horizontal = 14.dp, vertical = 10.dp),
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(10.dp)
                ) {
                    Icon(
                        imageVector = Icons.Filled.SportsEsports,
                        contentDescription = null,
                        tint = if (show) Color(0xFF00A5FF) else Color.White.copy(alpha = 0.7f),
                        modifier = Modifier.size(18.dp)
                    )
                    Text(
                        text = text,
                        color = Color.White.copy(alpha = 0.92f),
                        style = MaterialTheme.typography.bodyMedium.copy(
                            fontWeight = FontWeight.SemiBold,
                            shadow = Shadow(
                                color = Color.Black.copy(alpha = 0.7f),
                                offset = Offset(1f, 1f),
                                blurRadius = 3f
                            )
                        ),
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                        softWrap = false
                    )
                }
        }
    }

    /**
     * Центральный ряд: иконка Windows → стрелка → квадратная миниатюра обложки.
     * Если обложки нет — показываем стилизованный плейсхолдер.
     */
    @Composable
    private fun BootCenterRow(
        bitmap: Bitmap?,
        appId: Int?,
        activity: Activity
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(20.dp)
        ) {
            // Иконка Windows — статичная, без мигания
            Box(
                modifier = Modifier
                    .size(72.dp)
                    .shadow(elevation = 16.dp, shape = RoundedCornerShape(16.dp), ambientColor = Color(0xFF00A5FF), spotColor = Color(0xFF00A5FF))
                    .clip(RoundedCornerShape(16.dp))
                    .background(
                        Brush.linearGradient(
                            colors = listOf(Color(0xFF1B2233), Color(0xFF0B0F1A))
                        )
                    ),
                contentAlignment = Alignment.Center
            ) {
                Image(
                    painter = painterResource(R.drawable.ic_windows_10),
                    contentDescription = null,
                    modifier = Modifier.size(40.dp)
                )
            }

            // Стрелка перехода
            Box(
                modifier = Modifier.size(28.dp),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = Icons.AutoMirrored.Filled.ArrowForward,
                    contentDescription = null,
                    tint = Color(0xFF00A5FF),
                    modifier = Modifier.size(28.dp)
                )
            }

            // Миниатюра обложки (Steam hero / local bitmap / placeholder)
            Box(
                modifier = Modifier
                    .size(72.dp)
                    .shadow(elevation = 14.dp, shape = RoundedCornerShape(12.dp), ambientColor = Color.Black, spotColor = Color.Black)
                    .clip(RoundedCornerShape(12.dp))
                    .background(Color(0xFF1B2233))
            ) {
                if (bitmap != null) {
                    Image(
                        bitmap = bitmap.asImageBitmap(),
                        contentDescription = null,
                        contentScale = ContentScale.Crop,
                        modifier = Modifier.fillMaxSize()
                    )
                } else if (appId != null && appId > 0) {
                    val capsuleUrl = "https://shared.fastly.steamstatic.com/store_item_assets/steam/apps/$appId/capsule_231x87.jpg"
                    var capsuleFile by remember(appId) { mutableStateOf<java.io.File?>(SteamImageCache.getCachedFile(activity, capsuleUrl)) }
                    LaunchedEffect(appId) {
                        if (capsuleFile == null) {
                            capsuleFile = SteamImageCache.downloadIfNeeded(activity, capsuleUrl)
                        }
                    }
                    if (capsuleFile != null) {
                        AsyncImage(
                            model = capsuleFile,
                            contentDescription = null,
                            contentScale = ContentScale.Crop,
                            modifier = Modifier.fillMaxSize()
                        )
                    } else {
                        ThumbnailPlaceholder()
                    }
                } else {
                    ThumbnailPlaceholder()
                }
            }
        }
    }

    @Composable
    private fun ThumbnailPlaceholder() {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(
                    Brush.linearGradient(
                        colors = listOf(
                            Color(0xFF243049),
                            Color(0xFF111521)
                        )
                    )
                ),
            contentAlignment = Alignment.Center
        ) {
            Icon(
                imageVector = Icons.Filled.Computer,
                contentDescription = null,
                tint = Color.White.copy(alpha = 0.45f),
                modifier = Modifier.size(32.dp)
            )
        }
    }

    @Composable
    private fun LoadingBar() {
        LinearProgressIndicator(
            modifier = Modifier
                .width(220.dp)
                .height(3.dp)
                .clip(RoundedCornerShape(2.dp)),
            color = Color(0xFF00A5FF),
            trackColor = Color(0xFF1B2130)
        )
    }

    @Synchronized
    fun show(textResId: Int) {
        if (isShowing()) return
        close()
        if (dialog == null) create()
        exitTriggered = false
        stageText = activity.getString(textResId)
        dialog?.show()
        isVisible = true
    }

    fun showOnUiThread(textResId: Int) {
        activity.runOnUiThread { show(textResId) }
    }

    @Synchronized
    fun setTitle(title: CharSequence?) {
        this.titleText = title
    }

    @Synchronized
    fun setSteamAppId(steamAppId: String?) {
        this._steamAppId = steamAppId
    }

    @Synchronized
    fun setSubtitle(subtitle: CharSequence?) {
        this.subtitleText = subtitle
    }

    fun setSubtitleOnUiThread(subtitle: CharSequence?) {
        activity.runOnUiThread { setSubtitle(subtitle) }
    }

    @Synchronized
    fun setStage(stage: CharSequence?) {
        this.stageText = stage
    }

    fun setStageOnUiThread(stage: CharSequence?) {
        activity.runOnUiThread { setStage(stage) }
    }

    @Synchronized
    fun updateText(text: String?) {
        setStage(text)
    }

    fun updateTextOnUiThread(text: String?) {
        activity.runOnUiThread { updateText(text) }
    }

    @Synchronized
    fun setCoverArt(bitmap: Bitmap?) {
        this.coverArtBitmap = bitmap
    }

    fun setCoverArtOnUiThread(bitmap: Bitmap?) {
        activity.runOnUiThread { setCoverArt(bitmap) }
    }

    @Synchronized
    fun close() {
        if (dialog == null) return
        val d = dialog
        dialog = null
        isVisible = false
        try {
            d?.dismiss()
        } catch (ignored: Exception) {
        }
    }

    fun closeOnUiThread() {
        activity.runOnUiThread {
            exitTriggered = true
            close()
        }
    }

    fun isShowing(): Boolean {
        return dialog != null && dialog!!.isShowing
    }
}
